swarm/network/stream/snapshot_retrieval_test.go

```diff
@@ -132,14 +172,152 @@ var retrievalSimServiceMap = map[string]simulation.ServiceFunc{
+	//array where the generated chunk hashes will be stored
+	conf.hashes = make([]storage.Address, 0)
+
+	ctx, cancelSimRun := context.WithTimeout(context.Background(), 3*time.Minute)
+	defer cancelSimRun()
+
+	filename := fmt.Sprintf("testing/snapshot_%d.json", nodeCount)
+	err := sim.UploadSnapshot(ctx, filename)
+	if err != nil {
+		return err
+	}
+
+	log.Info("Starting simulation")
+
+	result := sim.Run(ctx, func(ctx context.Context, sim *simulation.Simulation) error {
+		nodeIDs := sim.UpNodeIDs()
+		// first iteration: create addresses
+		for _, n := range nodeIDs {
+			//get the kademlia overlay address from this ID
+			a := n.Bytes()
+			//append it to the array of all overlay addresses
+			conf.addrs = append(conf.addrs, a)
+			//the proximity calculation is on overlay addr,
+			//the p2p/simulations check func triggers on enode.ID,
+			//so we need to know which overlay addr maps to which nodeID
+			conf.addrToIDMap[string(a)] = n
+		}
+
+		// now create random chunks
+		chunks := storage.GenerateRandomChunks(int64(chunkSize), chunkCount)
+		for _, chunk := range chunks {
+			conf.hashes = append(conf.hashes, chunk.Address())
+		}
+
+		log.Debug("random chunks generated, mapping keys to nodes")
+
+		// map addresses to nodes
+		mapKeysToNodes(conf)
+
+		// second iteration: storing chunks at the peer whose
+		// overlay address is closest to a particular chunk's hash
+		log.Debug("storing every chunk at correspondent node store")
+		for _, id := range nodeIDs {
+			// for every chunk for this node (which are only indexes)...
+			for _, ch := range conf.idToChunksMap[id] {
+				item, ok := sim.NodeItem(id, bucketKeyStore)
+				if !ok {
+					return fmt.Errorf("Error accessing localstore")
+				}
+				lstore := item.(chunk.Store)
+				// ...get the actual chunk
+				for _, chnk := range chunks {
+					if bytes.Equal(chnk.Address(), conf.hashes[ch]) {
+						// ...and store it in the localstore
+						if _, err = lstore.Put(ctx, chunk.ModePutUpload, chnk); err != nil {
+							return err
+						}
+					}
+				}
+			}
+		}
+
+		// now try to retrieve every chunk from every node
+		log.Debug("starting retrieval")
+		cnt := 0
 
-The snapshot should have 'streamer' in its service list.
-*/
+		for _, id := range nodeIDs {
+			item, ok := sim.NodeItem(id, bucketKeyFileStore)
+			if !ok {
+				return fmt.Errorf("No filestore")
+			}
+			fileStore := item.(*storage.FileStore)
+			for _, chunk := range chunks {
+				reader, _ := fileStore.Retrieve(context.TODO(), chunk.Address())
+				content := make([]byte, chunkSize)
+				size, err := reader.Read(content)
+				//check chunk size and content
+				ok := true
+				if err != io.EOF {
+					log.Debug("Retrieve error", "err", err, "hash", chunk.Address(), "nodeId", id)
+					ok = false
+				}
+				if size != chunkSize {
+					log.Debug("size not equal chunkSize", "size", size, "hash", chunk.Address(), "nodeId", id)
+					ok = false
+				}
+				// skip chunk "metadata" for chunk.Data()
```

Comment:
remove comment

Reply:
Why? Do you know what ` chunk.Data()[8:]` is? What if someone is onboarded?